### ISSUE TITLE
no-invalid-this incorrectly flags prototype methods

### ISSUE THREAD

**Issue by jscharett** [palantir/tslint#3921]
If I create a method on the prototype, this rule will flag it.  Seem like ti should ignore methods on "class-like" objects.

Example:
```
class A {}
A.prototype.b = function() {
    this.c = 1; // this line is flagged as error
};
```
I use this behavior to do some aop behavior on a third-party plugin, but this rule flags it as invalid.

**Comment by giladgray** [comment]
try adding a `this` arg to the function:

```ts
A.prototype.b = function(this: A) {
    this.c = 1; // this line is flagged as error
};
```

https://www.typescriptlang.org/docs/handbook/functions.html#this-parameters
